src/app/clusters/microwave-oven-control-server/microwave-oven-control-server.cpp

```diff
@@ -0,0 +1,294 @@
+    return (ep >= kMicrowaveOvenControlInstanceTableSize ? nullptr : gInstanceTable[ep]);
+}
+
+/**
+ * Set operational state instance
+ */
+void SetOPInstance(EndpointId aEndpoint, OperationalState::Instance * aInstance)
+{
+    uint16_t ep = emberAfGetClusterServerEndpointIndex(aEndpoint, MicrowaveOvenControl::Id,
+                                                       EMBER_AF_MICROWAVE_OVEN_CONTROL_CLUSTER_SERVER_ENDPOINT_COUNT);
+    // if endpoint is found
+    if (ep < kMicrowaveOvenControlInstanceTableSize)
+    {
+        gInstanceTable[ep] = aInstance;
+    }
+}
+
+} // namespace MicrowaveOvenControl
+} // namespace Clusters
+} // namespace app
+} // namespace chip
+
+/**********************************************************
+ * Callbacks Implementation
+ *********************************************************/
+/**
+ * @brief Microwave Oven Control Cluster SetCookingParameters Command callback (from client)
+ */
+bool emberAfMicrowaveOvenControlClusterSetCookingParametersCallback(
+    chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
+    const chip::app::Clusters::MicrowaveOvenControl::Commands::SetCookingParameters::DecodableType & commandData)
+{
+    uint8_t opState;
+    EndpointId endpointId                                  = commandPath.mEndpointId;
+    EmberAfStatus emberAfStatus                            = EMBER_ZCL_STATUS_SUCCESS;
+    Status status                                          = Status::Success;
+    OperationalState::Instance * gOperationalStateInstance = nullptr;
+
+    // get Operational State instance and current state
+    gOperationalStateInstance = GetOPInstance(endpointId);
+    if (gOperationalStateInstance == nullptr)
+    {
+        ChipLogProgress(Zcl, "Server didn't create instance of Operational State");
+        status = Status::InvalidInState;
+        goto exit;
+    }
+
+    opState = gOperationalStateInstance->GetCurrentOperationalState();
+    if (opState == to_underlying(OperationalStateEnum::kStopped))
+    {
+        auto & CookMode     = commandData.cookMode;
+        auto & CookTime     = commandData.cookTime;
+        auto & PowerSetting = commandData.powerSetting;
+
+        if (CookMode.HasValue())
+        {
+            // TODO: set Microwave Oven cooking mode by CookMode.Value().
+        }
+        else
+        {
+            // TODO: set Microwave Oven cooking mode to normal mode.
+        }
+
+        if (CookTime.HasValue())
+        {
+            // set Microwave Oven cooking time by getting input value.
+            if (CookTime.Value() < kMinCookTime || CookTime.Value() > kMaxCookTime)
+            {
+                status = Status::InvalidCommand;
+                ChipLogError(Zcl, "Failed to set cookTime, cookTime value is out of range");
+                goto exit;
+            }
+            else
+            {
+                emberAfStatus = CookTime::Set(endpointId, CookTime.Value());
```

Comment:
This only works with Ember post attribute change callbacks. It seems like the entire cluster is fully based on attribute-store, which makes it harder to make functional products/tests, since the examples then have to have distributed post attribute change callback logic to implement the clusters logic. For example, I don't see how the time countdown is implemented to properly react to changes, and it's not clear how this will be tested/testable in an automated way.

Reply:
@tcarmelveilleux For the command `SetCookingParameters` has the requirement as below, so we put the range check in post attribute change callbacks can't match the requirement, just put them in command callback . Is it better?

> If the values of the fields sent with the command do not meet the constraints of any of the
> associated attributes (i.e. bearing the same name as the field or as described in the field
> description), the server SHALL respond with a response of INVALID_COMMAND and the values of
> the attributes SHALL remain unchanged.

Reply:
@tcarmelveilleux For the command SetCookingParameters has the requirement as below, so we put the range check in post attribute change callbacks can't match the requirement, just put them in command callback . Is it better?

If the values of the fields sent with the command do not meet the constraints of any of the
associated attributes (i.e. bearing the same name as the field or as described in the field
description), the server SHALL respond with a response of INVALID_COMMAND and the values of
the attributes SHALL remain unchanged.